File: webrtc/src/main/java/dev/onvoid/webrtc/media/audio/AudioConverter.java
/*
 * Copyright 2021 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc.media.audio;

import dev.onvoid.webrtc.internal.DisposableNativeObject;

/**
 * Audio format converter to remix and resample audio input data. This converter
 * operates on audio frames of 10 milliseconds. Each sample is assumed to be a
 * 16-bit PCM sample.
 *
 * @author Alex Andres
 */
public class AudioConverter extends DisposableNativeObject {

	private final int srcSamples;

	private final int dstSamples;

	private final int dstSamplesOut;


	/**
	 * Creates a new {@code AudioConverter} with specified sampling frequency
	 * and channel parameters.
	 *
	 * @param srcSampleRate The sampling frequency of the input signal.
	 * @param srcChannels   The number of audio channels of the input signal.
	 * @param dstSampleRate The sampling frequency of the output signal.
	 * @param dstChannels   The number of audio channels of the output signal.
	 */
	public AudioConverter(int srcSampleRate, int srcChannels, int dstSampleRate,
			int dstChannels) {
		this.srcSamples = srcSampleRate / 100 * srcChannels;
		this.dstSamples = Math.max(srcSampleRate, dstSampleRate) / 100 * dstChannels;
		this.dstSamplesOut = (dstSampleRate / 100) * dstChannels;

		initialize(srcSampleRate, srcChannels, dstSampleRate, dstChannels);
	}

	/**
	 * Calculates the buffer size in bytes for the destination buffer used in
	 * {@link #convert}.
	 *
	 * @return The target buffer size in bytes.
	 */
	public int getTargetBufferSize() {
		return dstSamples * 2;
	}

	/**
	 * Converts the input samples into the output sample buffer with the
	 * sampling frequency and channel layout specified in the constructor. The
	 * audio input must be of the length of 10 milliseconds. Accordingly, the
	 * output has the same length of 10 milliseconds.
	 *
	 * @param src The audio samples to convert.
	 * @param dst The output buffer for converted audio samples.
	 *
	 * @return The number of converted samples.
	 *
	 * @throws IllegalArgumentException if the buffer sizes do not match the
	 *                                  frame sizes.
	 */
	public int convert(byte[] src, byte[] dst) {
		if (src.length / 2 < srcSamples) {
			throw new IllegalArgumentException(String.format(
					"Insufficient samples input length: %d vs. %d",
					src.length / 2, srcSamples));
		}
		if (dst.length / 2 < dstSamples) {
			throw new IllegalArgumentException(String.format(
					"Insufficient samples output length: %d vs. %d",
					dst.length / 2, dstSamples));
		}

		convertInternal(src, srcSamples, dst, dstSamples);

		return dstSamplesOut;
	}

	@Override
	public native void dispose();

	private native void initialize(int srcSampleRate, int srcChannels,
			int dstSampleRate, int dstChannels);

	public native void convertInternal(byte[] src, int nSrcSamples, byte[] dst,
			int nDstSamples);

}

File: webrtc-jni/src/main/cpp/src/media/audio/AudioConverter.cpp
/*
 * Copyright 2021 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "media/audio/AudioConverter.h"

#include <memory>

#include "audio/utility/channel_mixer.h"
#include "common_audio/resampler/include/push_resampler.h"

namespace jni
{
    class CopyConverter : public AudioConverter {
        public:
            CopyConverter(size_t srcFrames, size_t srcChannels, size_t dstFrames, size_t dstChannels)
                : AudioConverter(srcFrames, srcChannels, dstFrames, dstChannels)
            {
            }

            ~CopyConverter() override
            {
            }

            void convert(const int16_t * src, size_t srcSize, int16_t * dst, size_t dstSize) override {
                checkSizes(srcSize, dstSize);

                if (src != dst) {
                    std::memcpy(dst, src, dstFrames * dstChannels * sizeof(int16_t));
                }
            }
    };


    class ChannelConverter : public AudioConverter {
        public:
            ChannelConverter(size_t srcFrames, size_t srcChannels, size_t dstFrames, size_t dstChannels)
                : AudioConverter(srcFrames, srcChannels, dstFrames, dstChannels)
            {
                webrtc::ChannelLayout srcLayout = webrtc::GuessChannelLayout(static_cast<int>(srcChannels));
                webrtc::ChannelLayout dstLayout = webrtc::GuessChannelLayout(static_cast<int>(dstChannels));

                mixer = std::make_unique<webrtc::ChannelMixer>(srcLayout, dstLayout);
                frame = std::make_unique<webrtc::AudioFrame>();

                frame->samples_per_channel_ = srcFrames;
                frame->num_channels_ = srcChannels;
            }

            ~ChannelConverter() override
            {
            }

            void convert(const int16_t * src, size_t srcSize, int16_t * dst, size_t dstSize) override {
                checkSizes(srcSize, dstSize);

                std::memcpy(frame->mutable_data(), src, srcFrames * srcChannels * sizeof(int16_t));

                mixer->Transform(frame.get());

                std::memcpy(dst, frame->data(), dstFrames * dstChannels * sizeof(int16_t));
            }

        private:
            std::unique_ptr<webrtc::ChannelMixer> mixer;
            std::unique_ptr<webrtc::AudioFrame> frame;
    };


    class ResampleConverter : public AudioConverter {
        public:
            ResampleConverter(size_t srcFrames, size_t srcChannels, size_t dstFrames, size_t dstChannels)
                : AudioConverter(srcFrames, srcChannels, dstFrames, dstChannels)
            {
                resampler = std::make_unique<webrtc::PushResampler<int16_t>>();
                resampler->InitializeIfNeeded(static_cast<int>(srcFrames * 100), static_cast<int>(dstFrames * 100), srcChannels);
            }

            ~ResampleConverter() override
            {
            }

            void convert(const int16_t * src, size_t srcSize, int16_t * dst, size_t dstSize) override {
                checkSizes(srcSize, dstSize);

                resampler->Resample(src, srcSize, dst, dstSize);
            }

        private:
            std::unique_ptr<webrtc::PushResampler<int16_t>> resampler;
    };


    class CompositionConverter : public AudioConverter {
        public:
            explicit CompositionConverter(std::vector<std::unique_ptr<AudioConverter>> converters_)
                : converters(std::move(converters_))
            {
                RTC_CHECK_GE(converters.size(), 2);

                // We need an intermediate buffer after every converter.
                for (auto it = converters.begin(); it != converters.end() - 1; ++it) {
                    buffers.push_back(std::vector<int16_t>((*it)->getDstFrames() * (*it)->getDstChannels(), 0));
                }
            }

            ~CompositionConverter() override
            {
            }

            void convert(const int16_t * src, size_t srcSize, int16_t * dst, size_t dstSize) override {
                converters.front()->convert(src, srcSize, buffers.front().data(), buffers.front().size());

                for (size_t i = 2; i < converters.size(); ++i) {
                    auto & src_buffer = buffers[i - 2];
                    auto & dst_buffer = buffers[i - 1];

                    converters[i]->convert(src_buffer.data(), src_buffer.size(), dst_buffer.data(), dst_buffer.size());
                }

                converters.back()->convert(buffers.back().data(), buffers.back().size(), dst, dstSize);
            }

        private:
            std::vector<std::unique_ptr<AudioConverter>> converters;
            std::vector<std::vector<int16_t>> buffers;
    };


	std::unique_ptr<AudioConverter> AudioConverter::create(size_t srcFrames, size_t srcChannels, size_t dstFrames, size_t dstChannels)
	{
        std::unique_ptr<AudioConverter> converter;

        if (srcChannels > dstChannels) {
            if (srcFrames != dstFrames) {
                std::vector<std::unique_ptr<AudioConverter>> converters;
                converters.push_back(std::unique_ptr<AudioConverter>(new ChannelConverter(srcFrames, srcChannels, srcFrames, dstChannels)));
                converters.push_back(std::unique_ptr<AudioConverter>(new ResampleConverter(srcFrames, dstChannels, dstFrames, dstChannels)));

                converter.reset(new CompositionConverter(std::move(converters)));
            }
            else {
                converter.reset(new ChannelConverter(srcFrames, srcChannels, dstFrames, dstChannels));
            }
        }
        else if (srcChannels < dstChannels) {
            if (srcFrames != dstFrames) {
                std::vector<std::unique_ptr<AudioConverter>> converters;
                converters.push_back(std::unique_ptr<AudioConverter>(new ResampleConverter(srcFrames, srcChannels, dstFrames, srcChannels)));
                converters.push_back(std::unique_ptr<AudioConverter>(new ChannelConverter(dstFrames, srcChannels, dstFrames, dstChannels)));

                converter.reset(new CompositionConverter(std::move(converters)));
            }
            else {
                converter.reset(new ChannelConverter(srcFrames, srcChannels, dstFrames, dstChannels));
            }
        }
        else if (srcFrames != dstFrames) {
            converter.reset(new ResampleConverter(srcFrames, srcChannels, dstFrames, dstChannels));
        }
        else {
            converter.reset(new CopyConverter(srcFrames, srcChannels, dstFrames, dstChannels));
        }

        return converter;
	}

    AudioConverter::AudioConverter() :
        srcFrames(0),
        srcChannels(0),
        dstFrames(0),
        dstChannels(0)
    {
    }

    AudioConverter::AudioConverter(size_t srcFrames, size_t srcChannels, size_t dstFrames, size_t dstChannels) :
        srcFrames(srcFrames),
        srcChannels(srcChannels),
        dstFrames(dstFrames),
        dstChannels(dstChannels)
    {
    }

    void AudioConverter::checkSizes(size_t srcSize, size_t dstCapacity) const {
        RTC_CHECK_EQ(srcSize, srcChannels * srcFrames);
        RTC_CHECK_GE(dstCapacity, dstChannels * dstFrames);
    }
}